### PR TITLE
BigAnimal: Preparing your GCP account

### PR DESCRIPTION
**Pull request by drothery-edb**
## What Changed?

https://enterprisedb.atlassian.net/browse/UPM-21148

### CODE REVIEW

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx [hunk 1]
```diff
@@ -1,3 +1,76 @@
...
+    | &lt;region> | Google Cloud region where your clusters are being deployed. See [Supported regions](../../overview/03a_region_support/#supported-google-cloud-regions) for a list of possible regions. |
+
+    Possible options are:
+
+    | Options | Description |
+    | ------- | ----------- |
+    | `-h` or `--help`| Displays the command help. |
+    | `-i` or `--instance-type` | Google Cloud instance type for the BigAnimal cluster. The help command provides a list of possible VM instance types. Choose the instance type that best suits your application and workload. Choose an instance type in the memory optimized M1, M2, or M3 series for large data sets. Choose from the compute-optimized C2 or C2D series for compute-bound applications. Choose from the accelerator-optimized A2 or A2 ultra for massively parallelized Compute Unified Device Architecture (CUDA) workloads. Choose from the general purpose M5 or M6I series if you don't require memory or compute optimization.|
+    | `-a` or `--high-availability` | *DEPRECATED* - Enables high availability for the cluster. See [Supported cluster types(../../overview/02_high_availability) for more information.|
```
Comment (valeriodelsarto) [inline product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:43]:
```suggestion
```

Comment (valeriodelsarto) [reply product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:43]: this option is not present in the script, was already deprecated for Azure and AWS and has not been added into the GCP one

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx [hunk 2]
```diff
@@ -1,3 +1,76 @@
...
+2.  From Google Cloud Shell, run the following command:
+
+    ```shell
+    curl -sL https://raw.githubusercontent.com/EnterpriseDB/cloud-utilities/main/gcp/biganimal-csp-preflight | bash -s <GCP-account-ID> <region> [options]
+    ```
+    The required arguments are:
+
+    | Argument | Description |
+    | -------- | ----------- |
+    | &lt;account-id> | Google Cloud account ID of your BigAnimal deployment. |
```
Comment (valeriodelsarto) [inline product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:34]:
```suggestion
    | <account-id> | Google Cloud account ID of your BigAnimal deployment. |
```

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx [hunk 3]
```diff
@@ -1,3 +1,76 @@
...
+    | &lt;region> | Google Cloud region where your clusters are being deployed. See [Supported regions](../../overview/03a_region_support/#supported-google-cloud-regions) for a list of possible regions. |
```
Comment (valeriodelsarto) [inline product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:35]:
```suggestion
    | <region> | Google Cloud region where your clusters are being deployed. See [Supported regions](../../overview/03a_region_support/#supported-google-cloud-regions) for a list of possible regions. |
```

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx [hunk 4]
```diff
@@ -1,3 +1,76 @@
...
+    | `-a` or `--high-availability` | *DEPRECATED* - Enables high availability for the cluster. See [Supported cluster types(../../overview/02_high_availability) for more information.|
+    | `-x` or `--cluster-architecture` | Defines the Cluster architecture and can be `single`, `ha`, or `eha`. See [Supported cluster types](/biganimal/release/overview/02_high_availability) for more information.|
+    | `-e` or `--endpoint` | Type of network endpoint for the BigAnimal cluster, either `public` or `private`. See [Cluster networking architecture](../creating_a_cluster/01_cluster_networking) for more information. |
+    | `-r` or `--activate-region` | Specifies region activation if no clusters currently exist in the region. |
+    | `--onboard` | Checks if the user and subscription are correctly configured.
+
+    The behavior of the script defaults to `--onboard` if you provide no other options.
+
+    For example, if you want to deploy a cluster in a Google Cloud account having an ID of `1234-5678-9012`, with an instance type of `M1`, in the `us-east-1` region, in a `public` endpoint, and with no existing cluster deployed, run the following command:
```
Comment (valeriodelsarto) [inline product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:51]:
```suggestion
    For example, if you want to deploy a cluster in a Google Cloud account having an ID of `1234-5678-9012`, with an instance type of `n2-standard-8`, in the `us-east1` region, with a `public` endpoint, and with no existing cluster deployed, run the following command:
```

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx [hunk 5]
```diff
@@ -1,3 +1,76 @@
...
+
+    ```shell
+    curl -sL https://raw.githubusercontent.com/EnterpriseDB/cloud-utilities/main/gcp/biganimal-csp-preflight | bash -s \
+    1234-5678-9012 \
+    us-east-1 \
+    --instance-type m1 \
```
Comment (valeriodelsarto) [inline product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:57]:
```suggestion
    --instance-type n2-standard-8 \
```

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx [hunk 6]
```diff
@@ -1,3 +1,76 @@
...
+    | `--onboard` | Checks if the user and subscription are correctly configured.
+
+    The behavior of the script defaults to `--onboard` if you provide no other options.
+
+    For example, if you want to deploy a cluster in a Google Cloud account having an ID of `1234-5678-9012`, with an instance type of `M1`, in the `us-east-1` region, in a `public` endpoint, and with no existing cluster deployed, run the following command:
+
+    ```shell
+    curl -sL https://raw.githubusercontent.com/EnterpriseDB/cloud-utilities/main/gcp/biganimal-csp-preflight | bash -s \
+    1234-5678-9012 \
+    us-east-1 \
```
Comment (valeriodelsarto) [inline product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:56]:
```suggestion
    us-east1 \
```

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx [hunk 7]
```diff
@@ -1,3 +1,76 @@
...
+    --onboard\
+    ```
+
+The script displays the following output:
+
+-   Whether your Google Cloud account restricts vCPUs, elastic IP addresses, VPCs, or NLBs in your region (and availability zone, if HA is enabled). 
+
+    ```
+    ##############################################
+    VALERIO, CAN YOU PROVIDE EXAMPLE OUTPUT, PLEASE?
```
Comment (valeriodelsarto) [inline product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:69]:
```suggestion
    Run GCP Preflight Checks with gcloud cli 436.0.0

    ####################################
    # Checking for enabled GCP APIs... #
    ####################################

    NAME                                     RESULT
    ---------------------------------------  ---------
    autoscaling.googleapis.com               Enabled
    cloudapis.googleapis.com                 Enabled
    compute.googleapis.com                   Enabled
    container.googleapis.com                 Enabled
    iam.googleapis.com                       Enabled
    iamcredentials.googleapis.com            Enabled
    run.googleapis.com                       Enabled
    secretmanager.googleapis.com             Enabled
    storage.googleapis.com                   Enabled
    vpcaccess.googleapis.com                 Enabled

    ##############################################
    Checking Service Quotas Limits on us-east1...
    ##############################################

    Resource              Quota Name                            Limit    Used     Required    Gap      Suggestion
    --------              ----------                            ----     -----    --------    ---      ----------
    n2-standard-2 vCPUs   N2_CPUS                               500      0        6           494      OK
    e2-standard-8 vCPUs   E2_CPUS                               2400     0        24          2376     OK
    Shared-Core vCPUs     CPUS                                  2400     0        2           2398     OK
    Static IP Addresses   STATIC_ADDRESSES                      700      0        1           699      OK
    VPCs                  NETWORKS                              50       14       1           35       OK
    Cloud Routers         ROUTERS                               20       6        1           13       OK

    #######################
    # Overall Suggestions #
    #######################

    Make sure the GCP Project ID 1234-5678-9012 is the one that you want to use for BigAnimal
    Make sure the GCP account used <YOUR_GCP_ACCOUNT> has rights to create custom roles, service accounts, keys and assign project grants

    Please use the Quotas page in the Google Cloud console if need to raise any service quota limits.
    See https://cloud.google.com/docs/quota_detail/view_manage#requesting_higher_quota for more info.
```

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx [hunk 8]
```diff
@@ -1,3 +1,76 @@
...
+    ```
+
+## Configure your Google Cloud account
+
+VALERIO, IS THERE SIMILAR GCP INFORMATION WE SHOULD SUPPLY?
+
+Open an AWS support request to remove restrictions for the resources with `NotAvailable` displayed in the `Suggestion` column. See [Request quota increase](https://docs.aws.amazon.com/servicequotas/latest/userguide/request-quota-increase.html). For default service quota limits in AWS, see [AWS service information](https://docs.aws.amazon.com/general/latest/gr/aws-general.pdf#aws-service-information).
```
Comment (valeriodelsarto) [inline product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:76]:
```suggestion
If needed, you can enable required APIs following this [GCP documentation page](https://cloud.google.com/service-usage/docs/enable-disable). You can also increase your quotas following this [Request a higher quota limit](https://cloud.google.com/docs/quota_detail/view_manage#requesting_higher_quota).
```

**Review by valeriodelsarto — COMMENTED** [review]
Adding a few comments, feel free to answer to my suggestions or ask for more info if needed! 🙏 
Let me know where when you need another review or final approval!

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx (hunk 2 above)
Comment (drothery-edb) [reply product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:34]: We need to use the code for the left angle brackets or we get an error when generating the output

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx (hunk 3 above)
Comment (drothery-edb) [reply product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:35]: required coding

**Comment by drothery-edb** [comment]
@valeriodelsarto thank you for your comments. Could you make the following edits to the script output?

```suggestion
    ######################################################
    Run GCP Preflight Checks with Google Cloud CLI 436.0.0

    Make sure the GCP Project ID 1234-5678-9012 is the one that you want to use for BigAnimal.
    Make sure the GCP account you are using has rights to create custom roles, service accounts, keys, and assign project grants.

    Use the Quotas page in the Google Cloud console if you need to raise any service quota limits.
    See https://cloud.google.com/docs/quota_detail/view_manage#requesting_higher_quota for more information.
```

**Comment by valeriodelsarto** [comment]
> @valeriodelsarto thank you for your comments. Could you make the following edits to the script output?
> 
> ```
>     ######################################################
>     Run GCP Preflight Checks with Google Cloud CLI 436.0.0
> 
>     Make sure the GCP Project ID 1234-5678-9012 is the one that you want to use for BigAnimal.
>     Make sure the GCP account you are using has rights to create custom roles, service accounts, keys, and assign project grants.
> 
>     Use the Quotas page in the Google Cloud console if you need to raise any service quota limits.
>     See https://cloud.google.com/docs/quota_detail/view_manage#requesting_higher_quota for more information.
> ```

I'm not sure I've understood what you are asking to change... are you asking to remove the GCP account name from the output? I've redacted mine in this page but in reality this is what the scrips shows (also the project ID name is one of ours):

```
...
Make sure the GCP Project ID development-data is the one that you want to use for BigAnimal
Make sure the GCP account used valerio.delsarto@enterprisedb.com has rights to create custom roles, service accounts, keys and assign project grants
...
```

**Comment by drothery-edb** [comment]
> > @valeriodelsarto thank you for your comments. Could you make the following edits to the script output?
> > ```
> >     ######################################################
> >     Run GCP Preflight Checks with Google Cloud CLI 436.0.0
> > 
> >     Make sure the GCP Project ID 1234-5678-9012 is the one that you want to use for BigAnimal.
> >     Make sure the GCP account you are using has rights to create custom roles, service accounts, keys, and assign project grants.
> > 
> >     Use the Quotas page in the Google Cloud console if you need to raise any service quota limits.
> >     See https://cloud.google.com/docs/quota_detail/view_manage#requesting_higher_quota for more information.
> > ```
> 
> I'm not sure I've understood what you are asking to change... are you asking to remove the GCP account name from the output? I've redacted mine in this page but in reality this is what the scrips shows (also the project ID name is one of ours):
> 
> ```
> ...
> Make sure the GCP Project ID development-data is the one that you want to use for BigAnimal
> Make sure the GCP account used valerio.delsarto@enterprisedb.com has rights to create custom roles, service accounts, keys and assign project grants
> ...
> ```

I [committed](https://github.com/EnterpriseDB/docs/pull/4302/commits/79488108b0c718dbadc604644db802d8ccaa063c) the copy edits I was suggesting for the actual script output. Is it possible to make those changes to the script output itself (other than the use of angle brackets around the project and account ids)?

**Comment by valeriodelsarto** [comment]
> I [committed](https://github.com/EnterpriseDB/docs/pull/4302/commits/79488108b0c718dbadc604644db802d8ccaa063c) the copy edits I was suggesting for the actual script output. Is it possible to make those changes to the script output itself (other than the use of angle brackets around the project and account ids)?

Sure! I'll create a PR on `cloud-utilities` as soon as I can and will keep you posted!

**Comment by valeriodelsarto** [comment]
PR created, will merge after a review https://github.com/EnterpriseDB/cloud-utilities/pull/124

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx [hunk 9]
```diff
@@ -1,3 +1,112 @@
...
+    Run GCP Preflight Checks with Google Cloud CLI 436.0.0
+
+    ####################################
+    # Checking for enabled GCP APIs... #
+    ####################################
+
+    NAME                                     RESULT
+    ---------------------------------------  ---------
+    autoscaling.googleapis.com               Enabled
+    cloudapis.googleapis.com                 Enabled
```
Comment (valeriodelsarto) [inline product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:75]:
```suggestion
    cloudapis.googleapis.com                 Enabled
    cloudresourcemanager.googleapis.com      Enabled
```

Comment (valeriodelsarto) [reply product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:75]: I forgot this before, is being added with https://github.com/EnterpriseDB/cloud-utilities/pull/124/files#diff-30c64f35caf3a8d9af1ebc2320b70ec40752f17fb2defbdd6d75be728e83ff87R509

File: product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx [hunk 10]
```diff
@@ -1,3 +1,129 @@
...
+- roles/iam.roleAdmin
+- roles/resourcemanager.projectIamAdmin
+- roles/compute.viewer
+
+Or an equivalent single role such as:
+- roles/owner
+
+## Required APIs and services
+
+Ensure the following Google Cloud  are enabled APIs:
```
Comment (valeriodelsarto) [inline product_docs/docs/biganimal/release/getting_started/preparing_cloud_account/preparing_gcp/index.mdx:21]:
```suggestion
Ensure the following Google Cloud APIs are enabled:
```

**Pull request merged**